cpp/src/arrow/flight/sql/server_session_middleware.cc

```diff
@@ -0,0 +1,179 @@
+    } else {
+      try {
+        const std::shared_lock<std::shared_mutex> l(session_store_lock_);
+        auto session = session_store_.at(session_id);
+        *middleware = std::shared_ptr<ServerSessionMiddleware>(
+            new ServerSessionMiddleware(this, incoming_headers,
+                                        session, session_id));
+      } catch (std::out_of_range& e) {
+        return Status::Invalid(
+            "Invalid or expired "
+            + static_cast<std::string>(kSessionCookieName) + " cookie.");
+      }
+    }
+
+    return Status::OK();
+  }
+
+  /// \brief Get a new, empty session option map and its id key.
+  std::shared_ptr<FlightSqlSession> GetNewSession(std::string* session_id) {
+    std::string new_id = boost::lexical_cast<std::string>(uuid_generator_());
+    *session_id = new_id;
+    auto session = std::make_shared<FlightSqlSession>();
+
+    const std::unique_lock<std::shared_mutex> l(session_store_lock_);
+    session_store_[new_id] = session;
+
+    return session;
+  }
+};
+
+ServerSessionMiddleware::ServerSessionMiddleware(ServerSessionMiddlewareFactory* factory,
+                                                 const CallHeaders& headers)
+    : factory_(factory), headers_(headers), existing_session(false) {}
+
+ServerSessionMiddleware::ServerSessionMiddleware(
+    ServerSessionMiddlewareFactory* factory, const CallHeaders& headers,
+    std::shared_ptr<FlightSqlSession> session,
+    std::string session_id)
+    : factory_(factory), headers_(headers), session_(session), existing_session(true) {}
+
+void ServerSessionMiddleware::SendingHeaders(AddCallHeaders* addCallHeaders) {
+  if (!existing_session && session_) {
+    addCallHeaders->AddHeader(
+        "set-cookie",
+        static_cast<std::string>(kSessionCookieName) + "=" + session_id_);
+  }
+}
+
+void ServerSessionMiddleware::CallCompleted(const Status&) {}
+
+bool ServerSessionMiddleware::HasSession() const {
+  return static_cast<bool>(session_);
+}
+std::shared_ptr<FlightSqlSession> ServerSessionMiddleware::GetSession() {
+  if (!session_)
+    session_ = factory_->GetNewSession(&session_id_);
+  return session_;
+}
+const CallHeaders& ServerSessionMiddleware::GetCallHeaders() const {
+  return headers_;
+}
+
+
+
+std::shared_ptr<ServerMiddlewareFactory> MakeServerSessionMiddlewareFactory() {
+  return std::shared_ptr<ServerSessionMiddlewareFactory>(
+      new ServerSessionMiddlewareFactory());
+}
+
+SessionOptionValue FlightSqlSession::GetSessionOption(const std::string_view k) {
+  const std::shared_lock<std::shared_mutex> l(map_lock_);
+  return map_.at(k);
+}
+void FlightSqlSession::SetSessionOption(const std::string_view, const SessionOptionValue& v) {
+  const std::unique_lock<std::shared_mutex> l(map_lock_);
+  map_[k] = v;
+}
+void FlightSqlSession::EraseSessionOption(const std::string_view) {
+  const std::unique_lock<std::shared_mutex> l(map_lock_);
+  map_.erase(k);
```

Comment:
What's the intended behavior if k is not found?

Reply:
>   /// \brief Idempotently remove key from this call's Session, if Session & key exist

NOOP.  Good to check this was sane; all good.